impl/src/main/java/org/jboss/weld/context/beanstore/AttributeBeanStore.java

```diff
@@ -160,7 +188,21 @@ protected NamingScheme getNamingScheme() {
     }
 
     public Iterator<BeanIdentifier> iterator() {
-        return beanStore.iterator();
+        Iterator<BeanIdentifier> iterator;
```

Comment:
> If (!isAttached() || !isAttributeLazyFetchingEnabled()) then return beanStore.iterator();

I think we can't use this. The iterator is used during context destruction performed during deactivation (see `AbstractManagedContext.deactivate()`) and after detaching (see `AbstractBoundContext.deactivate()`). So if using lazy fetching the `beanStore.iterator()` may not contain all bean identifiers (e.g. "unfetched" session attributes in case of session context).

On the other hand, this should be safe:

     if (isAttributeLazyFetchingEnabled()) {
            // Use on-demand transforming iterator for `getPrefixedAttributeNames()`
        } else {
            iterator = beanStore.iterator();
        }